server/session.py

```diff
@@ -521,9 +529,111 @@ def notify(self, height, touched):
+                        if delta >= int(mnstat[4]):
+                            mnstat.append(int(mnstat[4]))
+                        # use active seconds
+                        else:
+                            mnstat.append(delta)
+                    mn_queue.append(mnstat)
+            mn_queue = sorted(mn_queue, key=lambda x: x[8], reverse=True)
+            return mn_queue
+
+        def get_payment_position(payment_queue, address):
+            '''
+            Returns the position of the payment list for the given address.
+
+            payment_queue: position in the payment queue for the masternode.
+            address: masternode payee address.
+            '''
+            position = -1
+            for pos, mn in enumerate(payment_queue, start=1):
+                if mn[2] == address:
+                    position = pos
+                    break
+            return position
+
+        # Accordingly with the masternode payment queue, a custom list with 
+        # the masternode information including the payment position is returned.
+        if self.controller.cache_mn_height != self.height():
+            self.controller.cache_mn_height = self.height()
+            self.controller.mn_cache.clear()
+            full_mn_list = await self.daemon.masternode_list(['full'])
+            mn_payment_queue = get_masternode_payment_queue(full_mn_list)
+            mn_list = []
+            for key, value in full_mn_list.items():
+                mn_data = value.split()
+                mn_info = {}
+                mn_info['vin'] = key
+                mn_info['status'] = mn_data[0]
+                mn_info['protocol'] = mn_data[1]
+                mn_info['payee'] = mn_data[2]
+                mn_info['lastseen'] = mn_data[3]
+                mn_info['activeseconds'] = mn_data[4]
+                mn_info['lastpaidtime'] = mn_data[5]
+                mn_info['lastpaidblock'] = mn_data[6]
+                mn_info['ip'] = mn_data[7]
+                mn_info['paymentposition'] = get_payment_position(mn_payment_queue, mn_info['payee'])
+                balance = await self.controller.address_get_balance(mn_info['payee'])
+                mn_info['balance'] = sum(balance.values()) / self.controller.coin.VALUE_PER_COIN
+                mn_list.append(mn_info)
+            self.controller.mn_cache = mn_list
+
+        if payees:
+            if type(payees) is str:
+                result = [mn for mn in self.controller.mn_cache if mn['payee'] == payees]
+            else:
+                result = [mn for mn in self.controller.mn_cache for address in payees if mn['payee'] == address]
+        else:
+            result = self.controller.mn_cache
+        return result
+
+    async def masternode_info(self, address):
+        '''
+        Returns the full info of masternode.
+
+        address: masternode payee address.
+        '''
+        result = []
+        address_list = ''.join(address.split()).split(',')
```

Comment:
This will crash if address is not a string - you are using untrusted input in all these functions so all must be type checked for their operations.